compiler/noirc_frontend/src/monomorphization/tests.rs

```diff
@@ -0,0 +1,37 @@
+#![cfg(test)]
+use crate::tests::get_program;
+
+use super::{ast::Program, errors::MonomorphizationError, monomorphize};
+
+fn get_monomorphized(src: &str) -> Result<Program, MonomorphizationError> {
+    let (_parsed_module, mut context, errors) = get_program(src);
+    assert!(
+        errors.iter().all(|err| !err.is_error()),
+        "Expected monomorphized program to have no errors before monomorphization, but found: {errors:?}"
+    );
+
+    let crate_id = context.root_crate_id();
+    let main = context
+        .get_main_function(crate_id)
+        .unwrap_or_else(|| panic!("get_monomorphized: test program contains no 'main' function"));
+    monomorphize(main, &mut context.def_interner, false)
+}
+
+#[test]
+fn recursive_type_errors() {
+    let src = "
+        fn main() {
+            let _tree: Tree<Tree<Tree<()>>> = Tree::Branch(
+                Tree::Branch(Tree::Leaf, Tree::Leaf),
+                Tree::Branch(Tree::Leaf, Tree::Leaf),
+            );
+        }
+
+        enum Tree<T> {
```

Comment:
Will we get errors if we just have a recursive type defined but not instantiated anywhere? 
e.g. 
```noir
fn main() { ... do other stuff ... }
enum Tree {
   Branch(Tree),
   Leaf
}
```
Based off the PR I don't think we will.

Reply:
I also expect the error from enums to be the same as the one from structs.

I think we could unify this error with [ResolverError::SelfReferentialType](https://github.com/noir-lang/noir/blob/a5b47df4d7c1da746f5377f9c1d35b5a06ea5648/compiler/noirc_frontend/src/hir/resolution/errors.rs#L122). 
e.g. this struct
```noir
struct Tree {
    branch: Tree,
}
```
should trigger the same error about an infinitely sized type as the enum above

Reply:
> Will we get errors if we just have a recursive type defined but not instantiated anywhere?

We wouldn't. I'd have to add a separate check for it. The reason we couldn't only have it when the enum is defined is that we still need to catch types like `Tree<T>` in the original test you commented on which only become recursive when instantiated with `T = Tree<_>`.

Reply:
> We wouldn't. I'd have to add a separate check for it. 

Probably could have that check be a part of the checks that trigger [ResolverError::SelfReferentialType](https://github.com/noir-lang/noir/blob/a5b47df4d7c1da746f5377f9c1d35b5a06ea5648/compiler/noirc_frontend/src/hir/resolution/errors.rs#L122). I have not looked at the code though.